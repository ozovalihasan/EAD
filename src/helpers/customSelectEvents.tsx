export const handleMouseUpForSelect = (el: React.MutableRefObject<HTMLSelectElement | null>) => {
  el.current!.classList.toggle("hidden")
}

export const handleMouseLeaveForSelect = (el: React.MutableRefObject<HTMLSelectElement | null>) => {
  el.current!.classList.add("hidden")
}