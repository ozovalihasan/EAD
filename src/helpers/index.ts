export * from './setColorVariants';
export * from './customSelectEvents';
