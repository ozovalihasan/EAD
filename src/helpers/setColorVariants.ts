
import colors from "tailwindcss/colors"

export const availableColors = ["sky", "neutral", "green", "red", "yellow", "pink", "orange", "emerald", "violet"] as const;
export type availableColorsType = typeof availableColors[number];

const setColorVariants = (color?: availableColorsType) => {
  if (color) {
    localStorage.setItem("mainColor", JSON.stringify(color))
  } else {
    
    if (localStorage.getItem("mainColor")) {
      const colorFromLocalStorage = JSON.parse(localStorage.getItem("mainColor")!) as string
      if ( availableColors.includes( colorFromLocalStorage as availableColorsType) ) {
        color = colorFromLocalStorage as availableColorsType
      } else {
        color = availableColors[0]
        localStorage.setItem("mainColor", JSON.stringify(color))
      }
    } else {
      color = availableColors[0]
      localStorage.setItem("mainColor", JSON.stringify(color))
    }
  }

  // eslint-disable-next-line @typescript-eslint/non-nullable-type-assertion-style
  const root = document.querySelector('div#root') as HTMLDivElement;

  const colorPalette = colors[color];

  Object.entries(colorPalette).forEach( ( [num, value] ) => {
    root.style.setProperty(`--first-${num}`, value);  
  });
}

export { setColorVariants }