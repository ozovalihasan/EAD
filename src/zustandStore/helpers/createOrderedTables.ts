import { TablesType } from "@/zustandStore"

export const createOrderedTables = (tables: TablesType) => {
  return Object.keys(tables)
}