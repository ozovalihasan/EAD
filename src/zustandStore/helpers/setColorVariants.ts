
import colors from "tailwindcss/colors"

export const availableColors = ["neutral", "green", "sky", "red", "yellow", "pink", "orange", "emerald", "violet"] as const;
export type availableColorsType = typeof availableColors[number];

const setColorVariants = (color: availableColorsType) => {
  // eslint-disable-next-line @typescript-eslint/non-nullable-type-assertion-style
  const root = document.querySelector('div#root') as HTMLDivElement;

  localStorage.setItem("selected-color", color);
 
  const colorPalette = colors[color];

  Object.entries(colorPalette).forEach( ( [num, value] ) => {
    root.style.setProperty(`--first-${num}`, value);  
  });
}

export { setColorVariants }