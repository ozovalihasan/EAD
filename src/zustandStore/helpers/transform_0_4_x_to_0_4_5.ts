import { State } from "@/zustandStore/store";
import { TableValueType } from "@/zustandStore/tables";

export const transform_0_4_x_to_0_4_5 = (data: State) => {
  if (data.version == "0.4.5") {return data}
  
  data.version = "0.4.5";
  
  Object.values(data.tables).forEach((table: TableValueType) => {
    table.superclassId = ""
  })
  
  return data
}

export default transform_0_4_x_to_0_4_5;