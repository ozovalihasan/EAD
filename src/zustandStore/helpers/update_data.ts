import { State, throughEdgePartial } from "@/zustandStore";
import { TableValueType } from "@/zustandStore/tables";

export const update_data = (data: State) => {

  if ( !(["0.4.5", "0.4.6", "0.4.7"].includes( data.version )) ) {
    Object.values(data.tables).forEach((table: TableValueType) => {
      table.superclassId = ""
    })
  }

  if ( !(["0.4.7"].includes( data.version )) ) {
    
    data.edges = data.edges.map(
      edge => edge.type === throughEdgePartial.type ? edge : {...edge, data: {optional: false}} 
    )

    data.nodes = data.nodes.map(
      (node) => ({
        ...node, style: {width: node.width, height: node.height}
      })
    ) 

  }
  data.version = "0.4.7";
  
  return data;
}

export default update_data;