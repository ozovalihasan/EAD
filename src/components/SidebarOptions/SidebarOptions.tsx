import useStore from '@/zustandStore/store';
import { memo, useRef } from 'react';
import { handleMouseLeaveForSelect, handleMouseUpForSelect } from '@/helpers';

export interface SidebarOptionsType {
  tableId: string
}

export const SidebarOptions = memo(({tableId}: {tableId: string} ) => {

  const tables = useStore((state) => state.tables);
  const selectEl = useRef<HTMLSelectElement | null>(null);
  const changeTableSuperclass = useStore(store => store.changeTableSuperClass)
  
  return (
    <div 
      className='relative w-1/2 whitespace-nowrap'
      onMouseLeave={() => handleMouseLeaveForSelect(selectEl)}
    >
      <div 
        className='truncate p-2 custom-select-button rounded-none rounded-tr-md'
        onMouseUp={() => handleMouseUpForSelect(selectEl)} 
        title="Click to select a superclass to inherit"
      >
        {tables[tableId].superclassId === "" ? "Base" : `< ${tables[(tables[tableId].superclassId)].name}`}
      </div>
      <select
        ref={selectEl}
        className="
          hidden absolute left-0 top-full z-20 w-11/12 custom-select-options
          [&>option]:p-2 [&>option]:truncate
        "
        value={tables[tableId].superclassId}
        onChange={(event) => changeTableSuperclass(event, tableId)}
        title="Select a superclass to inherit"
        size={Object.keys(tables).length + 1 }
      >
        <option title={"ActiveRecord::Base"} value="" >ActiveRecord::Base</option>

        {Object.keys(tables).map((superTableId: string) => (
          <option
            key={superTableId}
            value={superTableId}
            disabled={superTableId == tableId}
            title={tables[superTableId].name + (superTableId == tableId ? "(Disabled)" : "")}
          >
            {tables[superTableId].name}
          </option>
        ))}

      </select>
    </div>
  )
})
