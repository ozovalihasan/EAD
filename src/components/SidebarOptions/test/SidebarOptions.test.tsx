import {  SidebarOptions } from '../SidebarOptions';
import { render, screen, renderHook, fireEvent } from "@testing-library/react";
import useStore from '@/zustandStore/store';
import { handleMouseLeaveForSelect, handleMouseUpForSelect } from '@/helpers';

jest.mock('@/helpers',  () => ({
  handleMouseUpForSelect: jest.fn(),
  handleMouseLeaveForSelect: jest.fn()
}))

let renderReadyComponent: JSX.Element;

beforeEach(() => {
  useStore.setState({
    tables: {
      "1": {name: "Mock Name", attributes: {}, superclassId: ""},
      "2": {name: "Mock Second Name", attributes: {}, superclassId: "1"},
      "3": {name: "Mock Third Name", attributes: {}, superclassId: "1"}
    },
    changeTableSuperClass: jest.fn()
  })
  
  renderReadyComponent = (
    <SidebarOptions tableId={"2"}></SidebarOptions>
  );
});


describe('<Sidebar />', () => {

  it('renders "Base" if it doesn"t exist', () => {

    render(<SidebarOptions tableId={"1"}></SidebarOptions> );

    expect(screen.getByText(/^Base$/i)).toBeInTheDocument();
  });

  it('renders the name of super class if it exists', () => {

    render(renderReadyComponent );

    expect(screen.getByText(/< Mock Name/)).toBeInTheDocument();
  });
  
  
  it('renders all necessary options', () => {

    render(renderReadyComponent );

    expect(screen.getByText(/^ActiveRecord::Base$/)).toBeInTheDocument();
    expect(screen.getByText(/^Mock Name$/)).toBeInTheDocument();
    expect(screen.getByText(/^Mock Second Name$/)).toBeInTheDocument();
    expect(screen.getByText(/^Mock Third Name$/)).toBeInTheDocument();
  });
  
  it('calls changeTableSuperClass function of the store when the selected option is changed', () => {

    render(renderReadyComponent );
    const { result } = renderHook(() => useStore());

    const selectEl = screen.getByTitle(/Select a superclass to inherit/)
    fireEvent.change( selectEl, {target: {value: '3'}})
    expect(result.current.changeTableSuperClass).toHaveBeenCalledTimes(1);
  });
  
  it('calls the handleMouseUpForSelect function', () => {

    render( renderReadyComponent );

    const buttonEl = screen.getByText(/< Mock Name/);

    fireEvent.mouseUp( buttonEl );
    
    expect(handleMouseUpForSelect).toHaveBeenCalledTimes(1)

  });
  
  it('calls the handleMouseLeaveForSelect function', () => {

    render( renderReadyComponent );

    const selectEl = screen.getByTitle(/Select a superclass to inherit/);

    fireEvent.mouseLeave( selectEl );

    expect(handleMouseLeaveForSelect).toHaveBeenCalledTimes(1)

  });


  it('renders correctly', () => {
    const renderedContainer = render(renderReadyComponent );
    expect(renderedContainer).toMatchSnapshot();
  });

});