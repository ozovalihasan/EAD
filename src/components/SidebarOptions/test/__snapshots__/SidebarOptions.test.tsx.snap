// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative w-1/2 whitespace-nowrap"
      >
        <div
          class="truncate p-2 custom-select-button rounded-none rounded-tr-md"
          title="Click to select a superclass to inherit"
        >
          &lt; Mock Name
        </div>
        <select
          class="
          hidden absolute left-0 top-full z-20 w-11/12 custom-select-options
          [&>option]:p-2 [&>option]:truncate
        "
          size="4"
          title="Select a superclass to inherit"
        >
          <option
            title="ActiveRecord::Base"
            value=""
          >
            ActiveRecord::Base
          </option>
          <option
            title="Mock Name"
            value="1"
          >
            Mock Name
          </option>
          <option
            disabled=""
            title="Mock Second Name(Disabled)"
            value="2"
          >
            Mock Second Name
          </option>
          <option
            title="Mock Third Name"
            value="3"
          >
            Mock Third Name
          </option>
        </select>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative w-1/2 whitespace-nowrap"
    >
      <div
        class="truncate p-2 custom-select-button rounded-none rounded-tr-md"
        title="Click to select a superclass to inherit"
      >
        &lt; Mock Name
      </div>
      <select
        class="
          hidden absolute left-0 top-full z-20 w-11/12 custom-select-options
          [&>option]:p-2 [&>option]:truncate
        "
        size="4"
        title="Select a superclass to inherit"
      >
        <option
          title="ActiveRecord::Base"
          value=""
        >
          ActiveRecord::Base
        </option>
        <option
          title="Mock Name"
          value="1"
        >
          Mock Name
        </option>
        <option
          disabled=""
          title="Mock Second Name(Disabled)"
          value="2"
        >
          Mock Second Name
        </option>
        <option
          title="Mock Third Name"
          value="3"
        >
          Mock Third Name
        </option>
      </select>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
